Source/DafnyLanguageServer/Workspace/DiagnosticPublisher.cs

```diff
@@ -26,11 +29,39 @@ public class DiagnosticPublisher : IDiagnosticPublisher {
     }
 
     private void PublishDocumentDiagnostics(DafnyDocument document) {
-      languageServer.TextDocument.PublishDiagnostics(new PublishDiagnosticsParams {
+      var diagnosticParameters = new PublishDiagnosticsParams {
         Uri = document.Uri,
         Version = document.Version,
         Diagnostics = GetDiagnostics(document).ToArray(),
-      });
+      };
+      languageServer.TextDocument.PublishDiagnostics(diagnosticParameters);
+    }
+
+    public void PublishVerificationDiagnostics(DafnyDocument document) {
+      if (document.LoadCanceled) {
+        // We leave the responsibility to shift the error locations to the LSP clients.
+        // Therefore, we do not republish the errors when the document (re-)load was canceled.
+        return;
+      }
+      // Document.GetDiagnostics() returns not only resolution errors, but previous verification errors
+      var currentDiagnostics =
+        document.Errors.GetDiagnostics(document.GetFilePath())
+          .Where(x => x.Severity == DiagnosticSeverity.Error).ToList();
+      var errors = currentDiagnostics
+        .Concat(document.OldVerificationDiagnostics)
+        .Where(x => x.Severity == DiagnosticSeverity.Error)
+        .ToArray();
+      var linesCount = Regex.Matches(document.Text.Text, "\r?\n").Count + 1;
```

Comment:
>I think that we want to support files with thousands of lines at some point, so we should refrain from doing operations that scale with the size of the file such as Regex.Matches(document.Text.Text, "\r?\n"). I think eventually we'll have an incremental parser that doesn't reparse the whole file on each change.

I mitigated the problem above by adding a property `LinesCount` to `DafnyDocument` which retrieves the number of lines from the range of the top verification tree, so that it is not computed on every notification, only when building the tree for the first time.